internal/clients/kafka/topic/topic.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		return err
+	}
+
+	t, ok := td[name]
+	if !ok {
+		return errors.New("no delete response for topic")
+	}
+	if t.Err != nil {
+		return errors.Wrap(t.Err, "cannot delete topic")
+	}
+
+	return nil
+}
+
+func Update(ctx context.Context, client *kadm.Client, desired *Topic) error {
+	// First Get existing Topic
+	existing, err := Get(ctx, client, desired.Name)
+	if err != nil {
+		return errors.Wrap(err, "cannot get topic")
+	}
+	if existing == nil {
+		return errors.New("topic does not exist")
+	}
+
+	_, err = client.UpdatePartitions(ctx, int(desired.Partitions), desired.Name)
+	if err != nil {
+		return errors.Wrap(err, "cannot update topic partitions")
+	}
+
+	if desired.Config != nil {
+		configs := desired.Config
+		for key, value := range configs {
+			s := kadm.AlterConfig{
+				Op:    kadm.SetConfig, // Op is the incremental alter operation to perform.
+				Name:  key,            // Name is the name of the config to alter.
+				Value: value,         // Value is the value to use when altering, if any.
+			}
+
+			r, err := client.AlterTopicConfigs(ctx, []kadm.AlterConfig{s}, desired.Name)
+			if err != nil {
+				return errors.Wrap(err, "cannot update topic configs")
+			}
+			if r[0].Err != nil {
+				return errors.Wrap(r[0].Err, "cannot update topic configs")
+			}
+		}
+	}
+
+	if desired.ReplicationFactor != existing.ReplicationFactor {
+		return errors.New("updating replication factor is not supported")
+	}
+
+	return nil
+}
+
+// Generate is used to convert Crossplane TopicParameters to Kafka's Topic.
+func Generate(name string, params *v1alpha1.TopicParameters) *Topic {
+	tpc := &Topic{
+		Name:              name,
+		ReplicationFactor: int16(params.ReplicationFactor),
+		Partitions:        int32(params.Partitions),
+	}
+
+	if len(params.Config) > 0 {
+		tpc.Config = make(map[string]*string, len(params.Config))
+		for k, v := range params.Config {
+			tpc.Config[k] = v
+		}
+	} else {
+		tpc.Config = nil
+	}
+
+	return tpc
+}
+
+// LateInitializeSpec fills empty spec fields with the data retrieved from Kafka.
+func LateInitializeSpec(params *v1alpha1.TopicParameters, observed *Topic) bool {
+	lateInitialized := false
+	if params.Config == nil {
+		params.Config = make(map[string]*string, len(observed.Config))
+	}
+
+	for k, v := range observed.Config {
+		if _, ok := params.Config[k]; !ok {
+			lateInitialized = true
+			params.Config[k] = v
+		}
+	}
+	return lateInitialized
+}
+
+// IsUpToDate returns true if the supplied Kubernetes resource differs from the
+// supplied Kafka Topic.
+func IsUpToDate(in *v1alpha1.TopicParameters, observed *Topic) bool {
+	if in.Partitions != int(observed.Partitions) {
+		return false
+	}
+	if in.ReplicationFactor != int(observed.ReplicationFactor) {
+		return false
+	}
+	if len(in.Config) != len(observed.Config) {
+		return false
+	}
+	for k, v := range observed.Config {
+		if iv, ok := in.Config[k]; !ok || iv != v {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
```suggestion
                if iv, ok := in.Config[k]; !ok || stringValue(iv) != stringValue(v) {
                        return false
                }
        }
        return true
 }

func stringValue(p *string) string {
       if p == nil {
               return ""
       }
       return *p
}
```

Otherwise, we are comparing pointers (`iv` and `v`) instead of their values.